### PR TITLE
PingPong test with TCP vs RDMA modes.

### PR DESCRIPTION
**Pull request by petro-rudenko**
PingPong test inspiried by [RDMA perftest](https://github.com/linux-rdma/perftest) ib_send_lat with 2 modes: RDMA, TCP. Parameters the same as in SendRecv benchmark tests, with new additional argument: `-m rdma|tcp`.

CC @yuvaldeg

### CODE REVIEW

File: src/test/java/com/ibm/disni/benchmarks/PingPongClient.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      recvBuf.clear();
+    }
+
+    printResults(startTime);
+    //close everything
+    endpoint.close();
+    endpointGroup.close();
+	System.exit(0);
+  }
+
```
Comment (patrickstuedi) [inline src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Can you explain me why we need an extra PingPong benchmark? It seems it does pretty much what the SendRecv benchmark is doing already.

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: SendRecv benchmark sends bulk of Send and Recv requests in parallel. PingPong sends one by one Send and Recv request (only when it got send it sends Recv). This makes 1:1 correspondence with TCP Socket benchmark.

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: It sends `queueDepth` number of send requests at the beginning. To compare aples:aples with TCP need whether to run SendRecv benchmark with queue depth = 1 or have `queueDepth` TCP sockets to server or send `queueDepth * bufferSize` bigger buffer in TCP :
```
actualRecvs = 64
actualSends = 64
opCount 64
actualRecvs = 1
actualSends = 1
opCount 65
actualRecvs = 1
actualSends = 1
opCount 66
actualRecvs = 3
actualSends = 3
opCount 69
actualRecvs = 5
actualSends = 5
opCount 74
actualRecvs = 4
actualSends = 4
opCount 78
```
In PingPong benchmark flow state is controlled by benchmark itself:
```
send PING
recv PONG
send PING
...
```

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Also PingPong allows to test on big buffers (100MB+) where RDMA shows good performance compared to TCP. With queueDepth  need to allocate multiple such buffers.

Comment (patrickstuedi) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Sorry, still confused, when you say "It sends queueDepth number of send requests at the beginning." you mean the new PingPong benchmark does that?

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: No PingPong is basically SendRecv with `queueDepth` = 1. Existing SendRecvClient has configuration parameter `queueDepth` which is 64 by default. This makes difficult to map to TCP world (64 connections or 64 bigger payload size, etc). To makes things simpler PingPong compares single qp with queueDepth = 1 with a single TCP socket pair. They both exchanges buffer of specified size. It doesn't designed to get full network bandwidth, just to rough estimate RDMA vs TCP latency on a single connection.

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: The [line you pointed](https://github.com/zrlio/disni/blob/master/src/test/java/com/ibm/disni/benchmarks/SendRecvClient.java#L77):
```java
int actualSends = endpoint.sendRequest(actualRecvs); // actualRecvs = 64 at the beginning

public int sendRequest(int res) throws IOException {
  int ready = Math.min(res, sendBudget); // res=64, sendBudget = 64
  for (int i = 0; i < ready; i++){ 
    int index = sendIndex % pipelineLength;
    sendIndex++;
    sendCall[index].execute(); // 64 times it calls this sendCall
    sendBudget--;
    sendPending++;
  }
  return ready;
 }
```

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: In PingPong `actualRecvs` and `actualSends` would be always 1

Comment (patrickstuedi) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Let me rephrase, my previous answer is misleading. The way SendRecv works from what I see is that if queuedepth is set to 1, it's doing exactly what PingPong. You can set queudepth to a larger value, then SendRecv will send a batch of requests at the beginning.

Comment (PepperJo) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Peter, I just had a look at the SendRecv benchmark code and with queue depth 1 it does exactly what your benchmark is supposed to do. The server initially waits for the client to send a request and sends on back once it received it from the client and so forth...

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Yes, the problem is that it's not fair to benchmark against TCP single socket with `queueDepth` > 1. That's why i've made separate benchmark to just compare RDMA vs TCP. Because user may forget to set `queueDepth` = 1 and compare single socket result with 64 queueDepth RDMA (even with 1 queueDepth RDMA showing very good result on 10Kb+ buffer sizes). Also to make iteration loop clearer to show that RDMA and TCP doing exactly the same thing (to not confuse with `actualRecv`, etc.).

Comment (patrickstuedi) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: If that's the issue then we just change the default value from 64 to 1. I do agree that the SendRecv code is a bit difficult to understand, but I don't think we want two benchmarks in the code doing exactly the same in a different way. We could replace the existing SendRecv benchmark with the PingPong benchmark, but then we would loose functionality (the SendRecv benchmark can work with different queue depths).

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: Ok, i'll add TCP part to SendRecv test. We'll tell our users that was interested in this functionality to run RDMA mode with `-q 1` when comparing to TCP

Comment (petro-rudenko) [reply src/test/java/com/ibm/disni/benchmarks/PingPongClient.java]: @patrickstuedi sure. Just updated existing tests with TCP mode.

**Comment by patrickstuedi** [comment]
Thinking about this a bit more.. What if we do the following? We keep the original version of the code, but rename the classes PingPong.. to something like RDMAvsTcpBenchmark and also change the default setting in the client so that it runs both the RDMA and the TCP version and prints out the performance for both. What do you think?

**Comment by petro-rudenko** [comment]
@patrickstuedi i've reverted back to add TCP case for SendRecv benchmark test. So should we return back to have separate class for this?

**Comment by patrickstuedi** [comment]
Just revert it back to the original PR with the PingPong benchmark that also has the TCP benchmark in it, but please rename the class, and also make sure it runs both tcp and rdma by default.

**Comment by petro-rudenko** [comment]
Done
